[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112837
* #112885
* #112836

As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

fegin committed:
[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict

As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

LucasLLC commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -77,9 +82,13 @@ class StateDictOptions:
-    - ``fsdp_state_dict_type``: if the model contains FSDP sharded submodules,
-      what FSDP state_dict type should be used.
-      The default value is SHARDED_STATE_DICT.
+    - ``full_state_dict``: if this is set to True, all the tensors in the
```
@fegin n00b question - Assuming `AsyncCollectiveTensors` are still returned without waiting?

wz337 replied:
I'll update the dtensor gathering to use the `full_tensor()` API in a follow-up PR, which will always return synchronously so we can make sure that the value in the state_dict is correct.

LucasLLC reviewed (approved):
Still catching up on the underlying implementations but LGTM!

wz337 commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1468,7 +1468,7 @@ def _unflatten_orig_param_states(
         elif not cpu_offload:
             with SimpleProfiler.profile("clone"):
-                value = value.detach.clone()
+                value = value.detach().clone()
```
lol. How come our tests haven't not caught this?

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][11/N] Implement cpu_offload and full_state_dict for get_state_dict"


As title

Differential Revision: [D50962991](https://our.internmc.facebook.com/intern/diff/D50962991/)

[ghstack-poisoned]

carmocca commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -77,9 +82,13 @@ class StateDictOptions:
+      returned state_dict will be gathered. No ShardedTensor and DTensor
+      will be in the returned state_dict.
+
+    - ``cpu_offload``: offload all the tensors to cpu. To prevent CPU OOM, if
+      ``full_state_dict`` is also true, then only the rank0 will get the
+      state_dict and all other ranks will get empty state_dict.
```
Hi! This is a sensible default, but why wasn't `rank0_only` exposed in this options class in addition to `cpu_offload`? For setups with enough RAM, loading the full CPU model weights on all ranks could be desirable.

Thank you!

fegin replied:
This is a subtle issue that not all users may be aware. So we would like to avoid OOM issues as possible. And in many use cases, when users do full_state_dict, only rank0 is going to save the state_dict. What's the use case for all the ranks to save the duplicated states?
